src/charm.py

```diff
@@ -115,10 +116,44 @@ def __init__(self, *args):
             self, jobs=[{"static_configs": [{"targets": [f"*:{PROMETHEUS_PORT}"]}]}]
         )
         self._logging = LogProxyConsumer(
-            self, relation_name="logging", log_files=LOG_PATHS, container_name="discourse"
+            self, relation_name="logging", log_files=LOG_PATHS, container_name=CONTAINER_NAME
         )
         self._grafana_dashboards = GrafanaDashboardProvider(self)
 
+    def _on_start(self, _: ops.StartEvent) -> None:
+        """Handle start event.
+
+        Args:
+            event: Event triggering the start event handler.
+        """
+        if not self._is_setup_completed():
+            self._set_up_discourse()
+        if self._are_relations_ready():
+            self._activate_charm()
+
+    def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
+        """Handle upgrade charm event.
+
+        Args:
+            event: Event triggering the upgrade charm event handler.
+        """
+        if not self._is_setup_completed():
+            self._set_up_discourse()
+        if self._are_relations_ready():
+            self._activate_charm()
+
+    def _on_discourse_pebble_ready(self, evt: ops.PebbleReadyEvent) -> None:
+        """Handle discourse pebble ready event.
+
+        Args:
+            event: Event triggering the discourse pebble ready event handler.
+        """
+        if not self._is_setup_completed():
+            self._set_up_discourse()
+        self._config_changed(evt)
```

Comment:
Generally speaking, the event handler functions should not call each other, instead the common logic of handling an event should be in a separate function which is used by both event handlers. That way it is easier to reason about how each event is handled and what logic is shared between events. For example, a developer may not expect changes to the `config-changed` event handler to have an impact on `pebble-ready`. Though it would be clear that if they changed shared logic for handling both events it would impact the handling of both events. Any thoughts on that?

Reply:
I've regrouped those blocs in one shared function